domain/access/state/permission.go

```diff
@@ -713,3 +722,192 @@ WHERE  grant_on = $M.grant_on
+			return apiUserUUID, nil
+		}
+	}
+
+	return apiUserUUID, accesserrors.PermissionNotValid
+}
+
+func (st *PermissionState) grantPermission(ctx context.Context, tx *sqlair.TX, args access.UpsertPermissionArgs) error {
+	userAccessLevel, err := st.userAccessLevel(ctx, tx, args.Subject, args.AccessSpec.Target)
+	if err != nil {
+		return errors.Annotatef(err, "getting current access for grant")
+	}
+	aSpec := corepermission.AccessSpec{
+		Target: args.AccessSpec.Target,
+		Access: userAccessLevel,
+	}
+	grantAccess := args.AccessSpec.Access
+	if aSpec.EqualOrGreaterThan(grantAccess) {
+		return errors.Errorf("user %q already has %q access or greater", args.Subject, grantAccess)
+	}
+	if err := st.updatePermission(ctx, tx, args.Subject, args.AccessSpec.Target.Key, string(grantAccess)); err != nil {
+		return errors.Annotatef(err, "updating current access during grant")
+	}
+	return nil
+}
+
+func (st *PermissionState) revokePermission(ctx context.Context, tx *sqlair.TX, args access.UpsertPermissionArgs) error {
+	newAccess := args.AccessSpec.RevokeAccess()
+	if newAccess == corepermission.NoAccess {
+		err := st.deletePermission(ctx, tx, args.Subject, args.AccessSpec.Target)
+		return errors.Annotatef(err, "revoking %q", args.AccessSpec.Access)
+	}
+	if err := st.updatePermission(ctx, tx, args.Subject, args.AccessSpec.Target.Key, string(newAccess)); err != nil {
+		return errors.Annotatef(err, "updating current access during revoke")
+	}
+	return nil
+}
+
+func (st *PermissionState) deletePermission(ctx context.Context, tx *sqlair.TX, subject string, target corepermission.ID) error {
+	type input struct {
+		Name    string `db:"name"`
+		GrantOn string `db:"grant_on"`
+	}
+
+	// The combination of grant_to and grant_on are guaranteed to be
+	// unique, thus it is all that is deleted to select the row to be
+	// deleted.
+	deletePermission := `
+DELETE
+FROM   permission
+WHERE  uuid IN (
+       SELECT p.uuid
+       FROM   permission p
+              LEFT JOIN user AS u ON u.uuid = p.grant_to
+       WHERE  u.name = $input.name
+       AND    p.grant_on = $input.grant_on
+)
+`
+	deletePermissionStmt, err := sqlair.Prepare(deletePermission, input{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	in := input{
+		Name:    subject,
+		GrantOn: target.Key,
+	}
+	err = tx.Query(ctx, deletePermissionStmt, in).Run()
+	if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return errors.Annotatef(err, "deleting permission of %q on %q", subject, target.Key)
+	}
+	return nil
+}
+
+func (st *PermissionState) updatePermission(ctx context.Context, tx *sqlair.TX, subjectName, grantOn, access string) error {
+	type input struct {
+		Name    string `db:"name"`
+		GrantOn string `db:"grant_on"`
+		Access  string `db:"access"`
+	}
+
+	updateQuery := `
+UPDATE permission
+SET    permission_type_id = (
+           SELECT id 
+           FROM   permission_access_type 
+           WHERE  type = $input.access
+       ) 
+FROM   permission p
```

Comment:
When you update `FROM`, one expects to see a value from that sub-statement used in one of the `SET`s.

This is simpler with the same clause as the delete, so:
```sql
UPDATE permission
SET    permission_type_id = (
           SELECT id 
           FROM   permission_access_type 
           WHERE  type = $input.access
       ) 
WHERE  grant_on = $input.grant_on
AND    grant_to IN (
    SELECT uuid 
    FROM   user
    WHERE  $input.name
    AND    removed = false
)
```